src/page/managers/LoginManager.ts

```diff
@@ -43,64 +51,72 @@ export default class LoginManager {
       return result;
   }
 
-  static async upsertUser(userData: Partial<UserData>): Promise<UserData> {
+  static async upsertUser(userData: Partial<UserData>, retry: number = 5): Promise<UserData> {
     logMethodCall("LoginManager.upsertUser", { userData });
+
+    if (retry === 0) {
+      throw new OneSignalError("Login: upsertUser failed: max retries reached");
```

Comment:
Do we have a catch for this so it doesn't bubble outside of the SDK? If we do catch some where, can we ensure that it doesn't create any side effects? That is does the SDK continue to operate as expected.

Reply:
We do have a **[catch here](https://github.com/OneSignal/OneSignal-Website-SDK/blob/user-model/retry-login/src/onesignal/OneSignal.ts#L160)**

I just added a commit to remove the outermost throw to prevent just that

Reply:
> create any side effects
Currently, it does create a side-effect: namely we never hydrate a `onesignal_id` so any operation dependent on that will fail.

We should consider what to do in this circumstance.

Reply:
Catching and rethrowing there is ok for now. Longer term the SDK will need to keep this operation in cache, so the operation isn't lost though.